distributed/active_memory_manager.py

```diff
@@ -131,23 +131,36 @@ def run_once(self, comm=None) -> None:
             for ts, (pending_repl, pending_drop) in self.pending.items():
                 if not ts.who_has:
                     continue
-                who_has = [ws_snd.address for ws_snd in ts.who_has - pending_drop]
+                who_has = {ws_snd.address for ws_snd in ts.who_has - pending_drop}
                 assert who_has  # Never drop the last replica
                 for ws_rec in pending_repl:
                     assert ws_rec not in ts.who_has
-                    repl_by_worker[ws_rec.address][ts.key] = who_has
+                    repl_by_worker[ws_rec.address][ts] = who_has
                 for ws in pending_drop:
                     assert ws in ts.who_has
-                    drop_by_worker[ws.address].add(ts.key)
+                    drop_by_worker[ws.address].add(ts)
 
             # Fire-and-forget enact recommendations from policies
-            # This is temporary code, waiting for
-            # https://github.com/dask/distributed/pull/5046
-            for addr, who_has in repl_by_worker.items():
-                asyncio.create_task(self.scheduler.gather_on_worker(addr, who_has))
-            for addr, keys in drop_by_worker.items():
-                asyncio.create_task(self.scheduler.delete_worker_data(addr, keys))
-            # End temporary code
+            stimulus_id = str(time())
+            for addr, ts_to_who_has in repl_by_worker.items():
+                self.scheduler.stream_comms[addr].send(
+                    {
+                        "op": "acquire-replicas",
+                        "keys": [ts.key for ts in ts_to_who_has],
+                        "stimulus_id": "acquire-replicas-" + stimulus_id,
```

Comment:
@fjetter unsure if, when a function generates as burst of messages, the stimulus ID is supposed to be unique to each message or not?

Reply:
TLDR Your code snippet aligns with my idea

We're up to define this how we want since this system is new. I'd like think of this as "one decision creates one ID". If one decision triggers multiple messages, I would probably put the same ID into the messages. My idea about this ID is to trace the impact of a decision. In this example, a `run_once` call is the stimulus/trigger and makes the decision to change something about the cluster state (this breaks into many smaller decisions but they have the same origin/root cause/base state). This specific `run_once` call will have an ID and we'll assign this ID to all messages generated by it.